### PR TITLE
bump terraform providers

### PR DESCRIPTION
**Pull request by flokli**
###### Motivation for this change
This updates the terraform provider update script (thanks @zimbatm for helping out here), and updates all providers.

I haven't finished rebuilding all providers yet.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/cluster/terraform-providers/data.nix [hunk 1]
```diff
@@ -600,13 +744,13 @@
       version = "1.5.0";
       sha256  = "0yhv9ahj6ajspgnl2f77gpyd6klq44dyl74lvl10bx6yy56abi2m";
     };
-  random =
```
Comment (zimbatm) [inline pkgs/applications/networking/cluster/terraform-providers/data.nix]: it looks like a bunch of providers are missing, mostly the old ones

**Comment by flokli** [comment]
`terraform-providers.google` and `terraform-providers.google-beta` fail to build:
```
❯ nix-build -A terraform-providers.google
these derivations will be built:
  /nix/store/ml84kwg898dfy9909cpnz5h428651d8q-terraform-provider-google-3.14.0.drv
building '/nix/store/ml84kwg898dfy9909cpnz5h428651d8q-terraform-provider-google-3.14.0.drv'...
unpacking sources
unpacking source archive /nix/store/jjxmsxl0haxj77xf2jd78cxwpaqp211i-source
source root is source
patching sources
configuring
building
go/src/github.com/terraform-providers/terraform-provider-google/google/config.go:51:2: code in directory /build/go/src/github.com/terraform-providers/terraform-provider-google/vendor/google.golang.org/api/sqladmin/v1beta4 expects import "google.golang.org/api/sql/v1beta4"
builder for '/nix/store/ml84kwg898dfy9909cpnz5h428651d8q-terraform-provider-google-3.14.0.drv' failed with exit code 1
error: build of '/nix/store/ml84kwg898dfy9909cpnz5h428651d8q-terraform-provider-google-3.14.0.drv' failed
❯ nix-build -A terraform-providers.google-beta
these derivations will be built:
  /nix/store/1l01hi54y4glw11hi9yb756ijkx4j8v0-terraform-provider-google-beta-3.14.0.drv
building '/nix/store/1l01hi54y4glw11hi9yb756ijkx4j8v0-terraform-provider-google-beta-3.14.0.drv'...
unpacking sources
unpacking source archive /nix/store/dij55sqlwxrpr98ak99w6vnw6m373qif-source
source root is source
patching sources
configuring
building
go/src/github.com/terraform-providers/terraform-provider-google-beta/google-beta/config.go:52:2: code in directory /build/go/src/github.com/terraform-providers/terraform-provider-google-beta/vendor/google.golang.org/api/sqladmin/v1beta4 expects import "google.golang.org/api/sql/v1beta4"
builder for '/nix/store/1l01hi54y4glw11hi9yb756ijkx4j8v0-terraform-provider-google-beta-3.14.0.drv' failed with exit code 1
error: build of '/nix/store/1l01hi54y4glw11hi9yb756ijkx4j8v0-terraform-provider-google-beta-3.14.0.drv' failed
```

There seems to be something wrong with the vendored libraries I guess?

File: pkgs/applications/networking/cluster/terraform-providers/data.nix (hunk 1 above)
Comment (flokli) [reply pkgs/applications/networking/cluster/terraform-providers/data.nix]: I invoked `hub api --paginate https://api.github.com/orgs/terraform-providers/repos | jq`, and couldn't find `random` etc in there. Is the GitHub API not showing some repos for some reason?

**Comment by flokli** [comment]
When checking out and building the provider manually, there seems to be some differing behaviour:

```
$ GO111MODULE=off go build
google/config.go:51:2: code in directory /home/zimbatm/go/src/github.com/terraform-providers/terraform-provider-google/vendor/google.golang.org/api/sqladmin/v1beta4 expects import "google.golang.org/api/sql/v1beta4"
```
(fails)

```
$ GO111MODULE=on go build -mod=vendor                                                                                       
```
(succeeds)

**Comment by flokli** [comment]
I opened a support ticket at GitHub about the fact that some repositories are missing from the API:

to reproduce, look for `terraform-provider-{dns,external,template}` in

```
hub api --paginate https://api.github.com/orgs/terraform-providers/repos | jq | vim -
```


Or if you want to go fancy, GraphQL:
```
$ hub api --paginate graphql -f query='
  query($endCursor: String) {
    repositoryOwner(login: "terraform-providers") {
      repositories(first: 100, after: $endCursor) {
        nodes {
          nameWithOwner
          name
          refs(first: 1, refPrefix: "refs/tags/", orderBy: {field: TAG_COMMIT_DATE, direction: DESC}) {
            nodes {
              name                                                                                                                                                                                                                                 
            }                                                                                                                                                                                                                                      
          }                                                                                                                                                                                                                                        
        }                                                                                                                                                                                                                                          
        pageInfo {                                                                                                                                                                                                                                 
          hasNextPage                                                                                                                                                                                                                              
          endCursor                                                                                                                                                                                                                                
        }                                                                                                                                                                                                                                          
      }                                                                                                                                                                                                                                            
    }                                                                                                                                                                                                                                              
  }                                                                                                                                                                                                                                                
' | jq | vim -
```

**Comment by flokli** [comment]
Alright, already heard back. That was quick! :-)

Turns out, these repos have been moved to the `hashicorp` organization, and we were fooled by the redirect - so we might want to look at the `hashicorp` organization too, and filter for `terraform-provider-*` repos to include these…

**Comment by flokli** [comment]
I pushed a new version, that uses `hub`, the GraphQL endpoint, and fetches from both `hashicorp` and `terraform-providers`.

This required some refactoring, but now we should have back all terraform providers and not run into GitHubs ratelimiting, as we're not generating a HTTP request for each provider.

PTAL.

File: pkgs/applications/networking/cluster/terraform-providers/update-all [hunk 2]
```diff
@@ -74,69 +91,81 @@ EOF
...
-Missing the GITHUB_AUTH env. This is required to work around the 60 request
-per hour rate-limit.
-
-Go to https://github.com/settings/tokens and create a new token with the
-"public_repo" scope.
-
-Then `export GITHUB_AUTH=<your user>:<your token>` and run this script again.
-HELP
-  exit 1
-fi
```
Comment (zimbatm) [inline pkgs/applications/networking/cluster/terraform-providers/update-all]: Is that not applicable anymore?

**Review by zimbatm — APPROVED** [review]

File: pkgs/applications/networking/cluster/terraform-providers/update-all (hunk 2 above)
Comment (flokli) [reply pkgs/applications/networking/cluster/terraform-providers/update-all]: No. `hub` has some other environment variables (see their manpage), and they shouldn't be necessary, given we only do ~10 http requests to their API.

**Comment by flokli** [comment]
I bumped the terraform providers once more, now including the fix for https://github.com/terraform-providers/terraform-provider-google/issues/5970.

File: pkgs/applications/networking/cluster/terraform-providers/data.nix [hunk 3]
```diff
@@ -32,37 +48,61 @@
...
       owner   = "terraform-providers";
       repo    = "terraform-provider-azuread";
-      rev     = "v0.7.0";
-      version = "0.7.0";
-      sha256  = "1a7w31dvjz5498445ia4m5gd1js3k7ghz6qqfq51f2n86iafs0xq";
+      rev     = "v0.8.0";
+      version = "0.8.0";
+      sha256  = "0vljhjbizxh5s8f2ki7yn6hzf5xbn5swhxmq9wpxmg7jw5z0k6ha";
+    };
+  azure-classic =
```
Comment (zimbatm) [inline pkgs/applications/networking/cluster/terraform-providers/data.nix]: it looks like the blacklist is not working anymore

Comment (flokli) [reply pkgs/applications/networking/cluster/terraform-providers/data.nix:91]: yeah, it didn't work anymore as the list of providers now contains the organization as well.
I prefixed all providers in the blacklist with the organization, now they get filtered out again.

In addition, we now support blacklisting from other organizations too :-)

**Comment by flokli** [comment]
Some providers that were moved to the `hashicorp` organization still have references to the `terraform-providers` organization, which needs patching.

**Comment by flokli** [comment]
I sent PRs to all of the providers that were renamed:

 - https://github.com/hashicorp/terraform-provider-archive/pull/67
 - https://github.com/hashicorp/terraform-provider-dns/pull/41
 - https://github.com/hashicorp/terraform-provider-external/pull/41
 - https://github.com/hashicorp/terraform-provider-http/pull/40
 - https://github.com/hashicorp/terraform-provider-local/pull/40
 - https://github.com/hashicorp/terraform-provider-null/pull/43
 - https://github.com/hashicorp/terraform-provider-random/pull/107
 - https://github.com/hashicorp/terraform-provider-template/pull/79
 - https://github.com/hashicorp/terraform-provider-tls/pull/71

and uploaded workarounds in `special-providers` for the time being.

I disabled the broken terraform providers
 - terraform-providers/terraform-provider-circonus
 - terraform-providers/terraform-provider-vmc

Apart from that, I could build `terraform-full` and `terraform_0_11-full`.

**Comment by flokli** [comment]
@GrahamcOfBorg build terraform-full terraform_0_11-full

**Comment by flokli** [comment]
@GrahamcOfBorg build terraform-full terraform_0_11-full

**Review by zimbatm — APPROVED** [review]

**Pull request merged**